File: src/main/java/uk/ac/napier/sa/model/RemoteDatabaseManager.java
package uk.ac.napier.sa.model;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import uk.ac.napier.sa.controller.adt.Product;

import java.sql.ResultSet;
import java.util.List;
import java.util.Map;

public sealed interface RemoteDatabaseManager permits DatabaseManager {


    boolean connect(@NotNull String host, int port, @NotNull String database, boolean useSSL, @NotNull String user, @NotNull String pass);

    boolean disconnect();

    boolean init(String path);

    boolean changePrice(int id, double price);

    boolean sell(int id, int saleType);

    boolean noStock();

    boolean purchase(int customerId, int productId);

    boolean checkLoyaltyCardEligibility(int id);

    boolean grantLoyalty(int id);

    void printLastNPurchases(int n);

    Product getProduct(int id);

    @Nullable ResultSet query(@NotNull String sql);

    @Nullable String retrieveProductName(int id);

    @NotNull List<Integer> stockMonitor();

    @NotNull Map<String, Object> generateReport();

}

File: src/main/java/uk/ac/napier/sa/model/RemoteDatabaseManager.kt
package uk.ac.napier.sa.model

import uk.ac.napier.sa.controller.adt.Product
import java.sql.ResultSet

interface RemoteDatabaseManager {

    fun connect(host: String, port: Int, database: String, useSSL: Boolean, user: String, pass: String): Boolean {
        return false
    }

    fun disconnect(): Boolean {
        return false
    }

    fun init(path: String?): Boolean {
        return false
    }

    fun changePrice(id: Int, price: Double): Boolean {
        return false
    }

    fun sell(id: Int, saleType: Int): Boolean {
        return false
    }

    fun noStock(): Boolean {
        return false
    }

    fun purchase(customerId: Int, productId: Int): Boolean {
        return false
    }

    fun checkLoyaltyCardEligibility(id: Int): Boolean {
        return false
    }

    fun grantLoyalty(id: Int): Boolean {
        return false
    }

    fun printLastNPurchases(n: Int)
    fun getProduct(id: Int): Product? {
        return null
    }

    fun query(sql: String): ResultSet? {
        return null
    }

    fun retrieveProductName(id: Int): String? {
        return ""
    }

    fun stockMonitor(): List<Int?> {
        return emptyList<Int>()
    }

    fun generateReport(): Map<String?, Any?> {
        return emptyMap<String?, Any>()
    }
}